Smt/Solver.lean

```diff
@@ -40,84 +42,197 @@ register_option smt.solver.path : String := {
-  ⟨s!"(declare-fun {quoteSymbol id} {s})" :: solver.commands⟩
+def declareFun (id : String) (s : Term) : SolverT m Unit := addCommand (.declare id s)
 
 /-- Define a function with name `id`, parameters `ps`, co-domain `s`,
     and body `t`. -/
 def defineFun (id : String) (ps : List (String × Term)) (s : Term) (t : Term) :
-  Solver :=
-  ⟨s!"(define-fun {quoteSymbol id} ({paramsToString ps}) {s} {t})" :: solver.commands⟩
-  where
-    paramsToString (ps : List (String × Term)) : String :=
-      String.intercalate " " (ps.map (fun (n, s) => s!"({quoteSymbol n} {s})"))
+  SolverT m Unit := addCommand (.defineFun id ps s t false)
 
 /-- Define a recursive function with name `id`, parameters `ps`, co-domain `s`,
     and body `t`. -/
-def defineFunRec (id : String) (ps : List (String × Term)) (s : Term)
-  (t : Term) :
-  Solver :=
-  ⟨s!"(define-fun-rec {quoteSymbol id} ({paramsToString ps}) {s} {t})" :: solver.commands⟩
-  where
-    paramsToString (ps : List (String × Term)) : String :=
-      String.intercalate " " (ps.map (fun (n, s) => s!"({quoteSymbol n} {s})"))
+def defineFunRec (id : String) (ps : List (String × Term)) (s : Term) (t : Term) :
+  SolverT m Unit := addCommand (.defineFun id ps s t true)
 
 /-- Assert that proposition `t` is true. -/
-def assert (t : Term) : Solver :=
-  ⟨s!"(assert {t})" :: solver.commands⟩
+def assert (t : Term) : SolverT m Unit := addCommand (.assert t)
 
 /-- Check if the query given so far is satisfiable and return the result. -/
-def checkSat (kind : Kind) (path : String) : IO String := do
-  let proc ← IO.Process.spawn {
-    stdin := IO.Process.Stdio.piped
-    stdout := IO.Process.Stdio.piped
-    stderr := IO.Process.Stdio.piped
-    cmd := path
-    args := match kind with
-      | Kind.cvc5    => #["-q"]
-      | Kind.z3      => #["-in"]
-      | Kind.vampire => #["--input_syntax", "smtlib2", "--output_mode", "smtcomp"]
-  }
-  let query := String.intercalate
-                 "\n" ("(exit)\n" :: "(check-sat)" :: solver.commands).reverse
-  proc.stdin.putStr query
-  -- Close stdin to signal EOF to the solver.
-  let (_, proc) ← proc.takeStdin
-  _ ← proc.wait
-  proc.stdout.readToEnd
+def checkSat : SolverT m Result := do
+  addCommand .checkSat
+  let out ← getSexp
+  let res ← match out with
+    | "sat"     => return .sat
+    | "unsat"   => return .unsat
+    | "unknown" => return .unknown
+    | _ => (throw (IO.userError s!"unexpected solver output: {repr out}") : IO Result)
+  return res
+
+/- TODO: We should probably parse the returned string into `Command`s or `String × Term`s. This is
+   bit tricky because we need to differentiate between literals and (user-defined) symbols. It
+   should be possible, however, because we store a list of all executed commands. -/
+/-- Return the model for a `sat` result. -/
+def getModel : SolverT m String := do
+  addCommand .getModel
+  -- TODO: Replace the lines below with `getSexp` once `Sexp.parse` becomes strict.
```

Comment:
`getSexp` depends on the parser being strict in rejecting incomplete S-exprs. But the current parser accepts incomplete ones like "(" and ")".

Reply:
Why does it depend on the strictness?

Reply:
I need to know when a solver is done responding to a command (e.g., `get-model`). Most solver responses are in s-expr form. So, one way to do so is to check if the number of opening parenthesis equals the number of closing parenthesis (this is what the code is currently doing), but that does not always work:
```smt2
(
(define-fun x () "(")
)
```
The above model is a valid s-expr but the number of opening and closing parenthesis is not equal. Instead of investing time writing a function that checks if an s-expr is complete, I decided to use your `Sexp.parse` (which I thought would reject incomplete s-exprs). But when I tried `Sexp.parse`, it accepted the first line in the above model as a valid s-expr. This caused `getSexp` to return before reading the remaining two lines (it's assuming that the solver is done responding to the cmmand) and `getModel` to return the empty model as a counter-example in the `smt` tactic.

So to answer your question, by "strict" I mean: reject incomplete/invalid s-exprs. I need the parser to be strict to know when a solver is done responding to a command.

Reply:
I see, I can make that change when rebasing.